flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SourceOperatorStreamTask.java

```diff
@@ -119,9 +113,13 @@ protected void finishTask() throws Exception {
     }
 
     @Override
-    public Future<Boolean> triggerCheckpointAsync(
+    public CompletableFuture<Boolean> triggerCheckpointAsync(
             CheckpointMetaData checkpointMetaData, CheckpointOptions checkpointOptions) {
         if (!isExternallyInducedSource) {
+            if (checkpointOptions.getCheckpointType().shouldDrain()) {
+                setSynchronousSavepoint(checkpointMetaData.getCheckpointId(), true);
+                mainMailboxExecutor.execute(this::endData, "Drain pipeline on stop-with-savepoint");
+            }
             return super.triggerCheckpointAsync(checkpointMetaData, checkpointOptions);
```

Comment:
I can't see any problem with it. I did spend some time thinking about calling the `endData` from within the `triggerCheckpointAsync`, so that we can call it in a single mailbox action. Unfortunately it is not possible without a bigger rework of said method. The PR already touches a lot of fragile code and I did not want to change even more.

At the same time I don't see a problem if there is something executed in between. The only important thing is that one is executed after the other. That's the only requirement.